src/main/java/ch/njol/skript/expressions/ExprItems.java

```diff
@@ -0,0 +1,172 @@
+import ch.njol.skript.lang.Literal;
+import ch.njol.skript.lang.SkriptParser.ParseResult;
+import ch.njol.skript.lang.util.SimpleExpression;
+import ch.njol.util.Kleenean;
+import ch.njol.util.coll.iterator.ArrayIterator;
+import ch.njol.util.coll.iterator.CheckedIterator;
+import ch.njol.util.coll.iterator.IteratorIterable;
+import org.bukkit.Material;
+import org.bukkit.event.Event;
+import org.bukkit.inventory.ItemStack;
+import org.eclipse.jdt.annotation.Nullable;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.List;
+import java.util.NoSuchElementException;
+
+@Name("Items")
+@Description("Items or blocks of a specific type, useful for looping.")
+@Examples({
+	"loop items of type ore and log:",
+	"\tblock contains loop-item",
+	"\tmessage \"Theres at least one %loop-item% in this block\"",
+	"drop all blocks at the player # drops one of every block at the player"
+})
+@Since("<i>unknown</i> (before 1.4.2)")
+public class ExprItems extends SimpleExpression<ItemStack> {
+
+	private static final ItemStack[] ALL_ITEMS = Arrays.stream(Material.values())
+		.map(ItemStack::new)
+		.toArray(ItemStack[]::new);
+
+	static {
+		Skript.registerExpression(ExprItems.class, ItemStack.class, ExpressionType.COMBINED,
+			"[(all [[of] the]|the|every)] block(s|[ ]type[s])",
+			"[(all [[of] the]|the|every)] blocks of type[s] %itemtypes%",
+			"[(all [[of] the]|the|every)] items of type[s] %itemtypes%");
+	}
+
+	@Nullable
+	private Expression<ItemType> itemTypeExpr;
+	private boolean items;
+	private ItemStack[] buffer = null;
+
+	@Override
+	@SuppressWarnings("unchecked")
+	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
+		items = matchedPattern == 2;
+		itemTypeExpr = matchedPattern == 0 ? null : (Expression<ItemType>) exprs[0];
+		if (itemTypeExpr instanceof Literal) {
+			for (ItemType itemType : ((Literal<ItemType>) itemTypeExpr).getAll())
+				itemType.setAll(true);
+		}
+		return true;
+	}
+
+	@Override
+	@Nullable
+	protected ItemStack[] get(Event event) {
+		if (buffer != null)
+			return buffer;
+		List<ItemStack> items = new ArrayList<>();
+		iterator(event).forEachRemaining(items::add);
+		if (itemTypeExpr instanceof Literal)
+			return buffer = items.toArray(new ItemStack[0]);
+		return items.toArray(new ItemStack[0]);
+	}
+
+	@Override
+	@Nullable
+	public Iterator<ItemStack> iterator(Event event) {
+		if (!items && itemTypeExpr == null)
+			return new Iterator<ItemStack>() {
+
+				private final Iterator<ItemStack> iterator = new ArrayIterator<>(Arrays.stream(ALL_ITEMS.clone())
+					.filter(itemStack -> itemStack.getType().isBlock())
+					.toArray(ItemStack[]::new));
+
+				@Override
+				public boolean hasNext() {
+					return iterator.hasNext();
+				}
+
+				@Override
+				public ItemStack next() {
+					return new ItemStack(iterator.next());
+				}
+
+				@Override
+				public void remove() {}
+			};
+
```

Comment:
```vb
Iterable<ItemStack>[] iterables = itemTypeExpr.stream(event).map(ItemType::getAll).toArray(Iterable[]::new);
if (items) {
  return Iterables.concat(iterables);
} else {
   return Iterators.filter(Iterables.concat(iterables), is -> is.getType().isBlock());
}
```
can we do something like this? `Iterables` is from `com.google.common.collect.Iterables`